### PR TITLE
KIP-546: Add Client Quota APIs

### PR DESCRIPTION
**Pull request by petedannemann**
KIP: https://cwiki.apache.org/confluence/display/KAFKA/KIP-546%3A+Add+Client+Quota+APIs+to+the+Admin+Client

API docs:
https://kafka.apache.org/protocol#The_Messages_DescribeClientQuotas

https://kafka.apache.org/protocol#The_Messages_AlterClientQuotas

I followed a similar code style to what was used for [alterconfigs](https://github.com/segmentio/kafka-go/blob/main/alterconfigs.go). Let me know if there is a better code style to use for this.

If this PR is too large to review I could split it into three pieces:
1. Adding support for float64 data types used in Kafka's protocol (quotas appear to be the only APIs that use this data type)
2. Add createquotas APIs
3. Add describequotas APIs

### CODE REVIEW

File: alterclientquotas.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+}
+
+type AlterClientQuotaOps struct {
+	// The quota configuration key.
+	Key string
+
+	// The quota configuration value to set, otherwise ignored if the value is to be removed.
+	Value float64
+
+	// Whether the quota configuration value should be removed, otherwise set.
```
Comment (hhahn-tw) [inline alterclientquotas.go:48]: > otherwise set.

Is there a word missing at the end there ("false" maybe)?

File: alterclientquotas_test.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+		entityType = "client-id"
+		entityName = "my-client-id"
+		key        = "producer_byte_rate"
+		value      = 500000.0
+	)
+
+	client, shutdown := newLocalClient()
+	defer shutdown()
+
+	expectedAlterResp := AlterClientQuotasResponse{
```
Comment (hhahn-tw) [inline alterclientquotas_test.go:27]: Maybe it's the standard practice to define the expected value early but it feels more logical to me just before the assert (and after the err check)

**Review by hhahn-tw — APPROVED** [review]

File: alterclientquotas.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+	Remove bool
+}
+
+type AlterClientQuotaResponseQuotas struct {
+	// The error code, or `0` if the quota alteration succeeded.
+	ErrorCode int16
```
Comment (rhansen2) [inline alterclientquotas.go:54]: In the client responses we usually wrap these details on an `error` typed field which can be constructed with https://github.com/segmentio/kafka-go/blob/5e7b724b8a23e2dc5adfb508c00dec5307719692/error.go#L658

File: protocol/describeclientquotas/describeclientquotas.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
+package describeclientquotas
+
+import "github.com/segmentio/kafka-go/protocol"
+
+func init() {
+	protocol.Register(&Request{}, &Response{})
+}
+
+type Request struct {
```
Comment (rhansen2) [inline protocol/describeclientquotas/describeclientquotas.go:9]: I believe you'll need a field to indicate v1 is a flexible message like
https://github.com/segmentio/kafka-go/blob/5e7b724b8a23e2dc5adfb508c00dec5307719692/protocol/addoffsetstotxn/addoffsetstotxn.go#L12

File: protocol/describeclientquotas/describeclientquotas.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+func (r *Request) ApiKey() protocol.ApiKey { return protocol.DescribeClientQuotas }
+
+func (r *Request) Broker(cluster protocol.Cluster) (protocol.Broker, error) {
+	return cluster.Brokers[cluster.Controller], nil
+}
+
+type Component struct {
+	EntityType string `kafka:"min=v0,max=v1"`
+	MatchType  int8   `kafka:"min=v0,max=v1"`
+	Match      string `kafka:"min=v0,max=v1,nullable"`
```
Comment (rhansen2) [inline protocol/describeclientquotas/describeclientquotas.go:23]: It looks like v1 is compact and nullable.

File: protocol/alterclientquotas/alterclientquotas.go [hunk 6]
```diff
@@ -0,0 +1,50 @@
+package alterclientquotas
+
+import "github.com/segmentio/kafka-go/protocol"
+
+func init() {
+	protocol.Register(&Request{}, &Response{})
+}
+
+// Detailed API definition: https://kafka.apache.org/protocol#The_Messages_AlterClientQuotas
+type Request struct {
```
Comment (rhansen2) [inline protocol/alterclientquotas/alterclientquotas.go:10]: I believe these also need a tagged field to indicated v1 is flexible. It looks like the string fields also became compact in v1.

File: describeclientquotas.go [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+)
+
+// DescribeClientQuotasRequest represents a request sent to a kafka broker to
+// describe client quotas.
+type DescribeClientQuotasRequest struct {
+	// Address of the kafka broker to send the request to
+	Addr net.Addr
+
+	// List of quota components to describe.
+	Components []DescribeClientQuotasRequestComponent `kafka:"min=v0,max=v1"`
```
Comment (rhansen2) [inline describeclientquotas.go:19]: I don't believe these kafka tags are needed for the client level requests/responses.

File: protocol/describeclientquotas/describeclientquotas.go [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+	Strict     bool        `kafka:"min=v0,max=v1"`
+}
+
+func (r *Request) ApiKey() protocol.ApiKey { return protocol.DescribeClientQuotas }
+
+func (r *Request) Broker(cluster protocol.Cluster) (protocol.Broker, error) {
+	return cluster.Brokers[cluster.Controller], nil
+}
+
+type Component struct {
```
Comment (rhansen2) [inline protocol/describeclientquotas/describeclientquotas.go:23]: I believe this also needs to be marked with a tag field for v1.

**Review by rhansen2 — APPROVED** [review]

**Pull request merged**